main.cpp

```diff
@@ -39,29 +39,251 @@ send me a DM to check your pull request
 
  Wait for my code review.
  */
-
+#include <iostream>
 /*
  copied UDT 1:
  */
+struct Filter
+{
+    double frequency;
+    int filterType;
+    float dryWet;
+
+    Filter() : frequency(4000.0), filterType(2), dryWet (1.0f) 
+    {
+        std::cout << "Spawning Filter object!\n" << std::endl;
+    }
+    ~Filter()
+    {
+        std::cout << "Destroying Filter object!\n" << std::endl;
+    }
+
+    struct FilterKnob
+    {
+        float knobPosition;
+        bool knobOverride;
+        int knobID;
+
+        FilterKnob()
+        {
+            knobPosition = 1.0f;
+            knobOverride = false;
+            knobID = 1;
+        }
 
+        void spawnKnob(int numberOfKnobs, float knobLocation);
+    };
+
+    void updateKnobPosition(FilterKnob knobOne, double nextKnobPosition);
+    double transitionToNextFrequency(Filter, double, double);
+};
+
+double Filter::transitionToNextFrequency(Filter thisFilter, double targetFrequency, double incrementAmount)
+{
+    std::cout << "\nThe starting frequency before transitioning is: " << thisFilter.frequency << "." << std::endl;
+
+    while(targetFrequency <= thisFilter.frequency || thisFilter.frequency <= targetFrequency)
+    {
+        std::cout << "Transitioning... frequency is: " << thisFilter.frequency << std::endl;
+
+        if(thisFilter.frequency < targetFrequency && targetFrequency - thisFilter.frequency > incrementAmount)
+        {
+            thisFilter.frequency += incrementAmount;
+        }
+        else if(thisFilter.frequency > targetFrequency && targetFrequency - thisFilter.frequency < incrementAmount)
+        {
+            thisFilter.frequency -= incrementAmount;
+        }
+        else
+        {
+            double remainder = targetFrequency - thisFilter.frequency;
+            thisFilter.frequency += remainder; // making sure we hit that final increment without exceeding target
+            
+            std::cout << "Frequency transition completed. Current frequency is: " << thisFilter.frequency << ".\n" << std::endl;
+            return thisFilter.frequency;
+        }  
+    }
+    
+    return thisFilter.frequency;
+}
+
+void Filter::FilterKnob::spawnKnob(int numberOfKnobs, float knobLocation)
+{
+    FilterKnob newKnob;
+
+    if(numberOfKnobs < 10) newKnob.spawnKnob(numberOfKnobs, knobLocation);
+    std::cout << "You are trying to spawn too many knobs at once!\n\n";
+}
+
+void Filter::updateKnobPosition(FilterKnob knobToUpdate, double nextKnobPosition)
+{
+    int knobID = knobToUpdate.knobID;
+    std::cout << "The knob to update is: " << knobID << ". The next knob position is: " << nextKnobPosition << ".\n" << std::endl;
+    std::cout << "Your filter frequency is:" << frequency << std::endl;
+}
```

Comment:
- [ ] add `FilterKnob` member variable to class. 
- [ ] remove `FilterKnob knobToUpdate` from parameter list
- [ ] revise function to use member variables of the calling instance.